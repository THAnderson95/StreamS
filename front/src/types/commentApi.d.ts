import { ATTACHMENT_TYPES } from "@/constants";

export interface APIDataCommentResponse {
  data: Array<CommentData>;
  paging: Paging;
}

interface CommentData {
  message: string;
  from: FromData;
  reactions?: Reactions;
  attachment?: Attachment;
  id: string;
  like_count: number;
  parent: Parent;
}

interface Parent {
  id: string;
}

interface FromData {
  id: string;
  name: string;
  picture: Picture;
}

interface Picture {
  data: PictureData;
}

interface PictureData {
  height: number;
  width: number;
  is_silhouette: boolean;
  url: string;
}

interface Attachment {
  media: {
    image: {
      height: number;
      src: string;
      width: number;
    };
    target: {
      id: string;
      url: string;
    };
    source?: string;
  };
  type: ATTACHMENT_TYPES;
  url: string;
}

interface Reactions {
  data: Array<Reaction>;
  paging: Paging;
}

interface Reaction {
  id: string;
  name: string;
  type: string;
  paging: Paging;
}

interface Paging {
  cursors: {
    before: string;
    after: string;
  };
}
