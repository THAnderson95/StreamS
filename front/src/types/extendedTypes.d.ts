import { CommentData } from "@api/commentApi";

export interface CommentDataWithReplies extends CommentData {
  replies?: Array<CommentData>;
}
