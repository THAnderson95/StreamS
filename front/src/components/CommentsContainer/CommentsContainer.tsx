import { CommentData } from "@/types";

import Comment from "@/components/Comment/Comment";

interface CommentContainerProps {
  data: Array<CommentData>;
}

export default function CommentsContainer({ data }: CommentContainerProps) {
  return (
    <div className="comments-container flex flex-col">
      {data.length > 0 && data.map((comment) => <Comment key={comment.id} data={comment} />)}
    </div>
  );
}
