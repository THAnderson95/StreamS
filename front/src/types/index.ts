export * from './commentApi';
export * from './extendedTypes'