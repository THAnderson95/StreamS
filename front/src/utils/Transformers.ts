import {
  APIDataCommentResponse,
  CommentData,
  CommentDataWithReplies,
} from "@/types";

export function addRepliesToComments(
  data: APIDataCommentResponse
): Array<CommentDataWithReplies> {
  // filter the comments that are child comments and convert them to `CommentDataWithReplies`
  const filteredComments: Array<CommentDataWithReplies> = data.data
    .filter((comment) => !comment.parent)
    .map((comment) => ({ ...comment, replies: [] }));

  // filter ONLY child replies
  const childReplies: Array<CommentData> = data.data.filter(
    (comment) => comment.parent
  );

  // merge them
  const addRepliesToComments: Array<CommentDataWithReplies> =
    filteredComments.map((parentComment) => {
      parentComment.replies = childReplies.filter(
        (replyComment) => replyComment.parent?.id === parentComment.id
      );
      return parentComment;
    });

  return addRepliesToComments;
}
