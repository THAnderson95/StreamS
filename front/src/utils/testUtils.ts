import { CommentDataWithReplies } from "@/types/extendedTypes";
import { ATTACHMENT_TYPES } from "@/constants";

function generateCommentWithoutParentsAndReplies(): CommentDataWithReplies {
  return {
    id: "1",
    message: "This is a comment without parents and without replies",
    from: {
      id: "user1",
      name: "User One",
      picture: {
        data: {
          height: 50,
          width: 50,
          is_silhouette: false,
          url: "https://example.com/user1.jpg",
        },
      },
    },
    like_count: 10,
    replies: [],
  };
}

function generateCommentWithParentAndReplies(): CommentDataWithReplies {
  return {
    id: "2",
    message: "This is a comment with a parent and with replies",
    from: {
      id: "user2",
      name: "User Two",
      picture: {
        data: {
          height: 50,
          width: 50,
          is_silhouette: false,
          url: "https://example.com/user2.jpg",
        },
      },
    },
    like_count: 5,
    parent: { id: "1" },
    attachment: undefined,
    reactions: undefined,
    replies: [
      {
        id: "3",
        message: "This is a reply to comment 2",
        from: {
          id: "user3",
          name: "User Three",
          picture: {
            data: {
              height: 50,
              width: 50,
              is_silhouette: false,
              url: "https://example.com/user3.jpg",
            },
          },
        },
        like_count: 2,
        parent: { id: "2" },
      },
    ],
  };
}

function generateCommentWithMultipleReactions(): CommentDataWithReplies {
  return {
    id: "4",
    message: "This is a comment with multiple reactions",
    from: {
      id: "user4",
      name: "User Four",
      picture: {
        data: {
          height: 50,
          width: 50,
          is_silhouette: false,
          url: "https://example.com/user4.jpg",
        },
      },
    },
    like_count: 20,
    reactions: {
      data: [
        { id: "reaction1", name: "User Five", type: "LIKE" },
        { id: "reaction2", name: "User Six", type: "LOVE" },
      ],
      paging: { cursors: { before: "", after: "" } },
    },
    replies: [],
  };
}

function generateCommentWithAttachment(): CommentDataWithReplies {
  return {
    id: "5",
    message: "",
    from: {
      id: "user5",
      name: "User Five",
      picture: {
        data: {
          height: 50,
          width: 50,
          is_silhouette: false,
          url: "https://example.com/user5.jpg",
        },
      },
    },
    like_count: 0,
    attachment: {
      media: {
        image: {
          height: 100,
          src: "https://example.com/attachment.jpg",
          width: 100,
        },
        target: { id: "attachment1", url: "https://example.com/attachment" },
      },
      type: ATTACHMENT_TYPES.PHOTO,
      url: "https://example.com/attachment",
    },
    replies: [],
  };
}

function generateCommentWithOnlyMessage(): CommentDataWithReplies {
  return {
    id: "6",
    message: "This is a comment with only a message",
    from: {
      id: "user6",
      name: "User Six",
      picture: {
        data: {
          height: 50,
          width: 50,
          is_silhouette: false,
          url: "https://example.com/user6.jpg",
        },
      },
    },
    like_count: 0,
    replies: [],
  };
}

export function generateCommentData(): Array<CommentDataWithReplies> {
  return [
    generateCommentWithoutParentsAndReplies(),
    generateCommentWithParentAndReplies(),
    generateCommentWithMultipleReactions(),
    generateCommentWithAttachment(),
    generateCommentWithOnlyMessage(),
  ];
}