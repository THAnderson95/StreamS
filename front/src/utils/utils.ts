import { REACTION_TYPE_SYMBOLS, REACTION_TYPES } from "@/constants";

export function getReactionSymbol(reaction: REACTION_TYPES): string {
    return REACTION_TYPE_SYMBOLS[reaction]
}