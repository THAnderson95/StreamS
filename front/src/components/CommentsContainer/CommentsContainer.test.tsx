import { act } from "react";

import { render } from "@testing-library/react";

import CommentsContainer from "./CommentsContainer";
import { generateCommentData } from "@/utils/testUtils";
import { CommentDataWithReplies } from "@/types/extendedTypes";

it("renders correctly and matches snapshot", async () => {
  const commentData: Array<CommentDataWithReplies> = generateCommentData();
  const { asFragment } = render(<CommentsContainer data={commentData} />);

  /* eslint-disable testing-library/no-unnecessary-act */
  await act(async () => {});
  // from the console: `ReactDOMTestUtils.act` is deprecated in favor of `React.act`... but it's importing from react ðŸ¤”

  expect(asFragment()).toMatchSnapshot();
});
