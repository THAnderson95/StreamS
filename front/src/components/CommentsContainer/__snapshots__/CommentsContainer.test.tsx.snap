// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly and matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="comments-container flex flex-col"
  >
    <div
      class="comment-container border border-2 flex flex-row p-2 m-2 bg-gray-300 rounded-2xl"
    >
      <div
        class="comment-picture"
      >
        <img
          alt="User One"
          class="comment-profile-image mr-4 rounded-full max-w-[50px] max-h-[50px]"
          src="https://example.com/user1.jpg"
        />
      </div>
      <div
        class="comment-info flex-1"
      >
        <div
          class="content-container bg-gray-200 rounded-lg m-2 p-2"
        >
          <div
            class="comment-name font-bold text-lg"
          >
            User One
          </div>
          <div
            class="comment-message"
          >
            This is a comment without parents and without replies
          </div>
        </div>
      </div>
    </div>
    <div
      class="comment-container border border-2 flex flex-row p-2 m-2 bg-gray-300 rounded-2xl"
    >
      <div
        class="comment-picture"
      >
        <img
          alt="User Two"
          class="comment-profile-image mr-4 rounded-full max-w-[50px] max-h-[50px]"
          src="https://example.com/user2.jpg"
        />
      </div>
      <div
        class="comment-info flex-1"
      >
        <div
          class="content-container bg-gray-200 rounded-lg m-2 p-2"
        >
          <div
            class="comment-name font-bold text-lg"
          >
            User Two
          </div>
          <div
            class="comment-message"
          >
            This is a comment with a parent and with replies
          </div>
        </div>
        <button
          class="show-button bg-blue-500 hover:bg-blue-600 text-white px-2 py-1 m-1 rounded"
        >
          Show Replies...
        </button>
      </div>
    </div>
    <div
      class="comment-container border border-2 flex flex-row p-2 m-2 bg-gray-300 rounded-2xl"
    >
      <div
        class="comment-picture"
      >
        <img
          alt="User Four"
          class="comment-profile-image mr-4 rounded-full max-w-[50px] max-h-[50px]"
          src="https://example.com/user4.jpg"
        />
      </div>
      <div
        class="comment-info flex-1"
      >
        <div
          class="content-container bg-gray-200 rounded-lg m-2 p-2"
        >
          <div
            class="comment-name font-bold text-lg"
          >
            User Four
          </div>
          <div
            class="comment-message"
          >
            This is a comment with multiple reactions
          </div>
        </div>
        <div
          class="comment-reactions"
        >
          <div
            class="comment-reactions flex flex-row"
          >
            <div
              class="reaction text-white  p-1 bg-gray-500 m-1 rounded-full w-auto min-w-1 h-8 flex items-center justify-center"
            >
              üëç
            </div>
            <div
              class="reaction text-white  p-1 bg-gray-500 m-1 rounded-full w-auto min-w-1 h-8 flex items-center justify-center"
            >
              ‚ù§Ô∏è
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="comment-container border border-2 flex flex-row p-2 m-2 bg-gray-300 rounded-2xl"
    >
      <div
        class="comment-picture"
      >
        <img
          alt="User Five"
          class="comment-profile-image mr-4 rounded-full max-w-[50px] max-h-[50px]"
          src="https://example.com/user5.jpg"
        />
      </div>
      <div
        class="comment-info flex-1"
      >
        <div
          class="content-container bg-gray-200 rounded-lg m-2 p-2"
        >
          <div
            class="comment-name font-bold text-lg"
          >
            User Five
          </div>
          <div
            class="comment-message"
          />
          <div
            class="comment-attachment"
          >
            <img
              class="attachment-photo p-2"
              src="https://example.com/attachment.jpg"
            />
          </div>
        </div>
      </div>
    </div>
    <div
      class="comment-container border border-2 flex flex-row p-2 m-2 bg-gray-300 rounded-2xl"
    >
      <div
        class="comment-picture"
      >
        <img
          alt="User Six"
          class="comment-profile-image mr-4 rounded-full max-w-[50px] max-h-[50px]"
          src="https://example.com/user6.jpg"
        />
      </div>
      <div
        class="comment-info flex-1"
      >
        <div
          class="content-container bg-gray-200 rounded-lg m-2 p-2"
        >
          <div
            class="comment-name font-bold text-lg"
          >
            User Six
          </div>
          <div
            class="comment-message"
          >
            This is a comment with only a message
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
